tiledb/api/src/column.rs

```diff
@@ -0,0 +1,416 @@
+use crate::error::Error;
+use crate::Datatype;
+use crate::Result as TileDBResult;
+
+/// The ReferenceColumn is used when we can safely use the existing Vec's or
+/// slices being used to create a column. This avoids the allocation, copy, and
+/// eventual deallocation when providing column data to TileDB.
+pub struct ReferenceColumn<'data> {
+    num_values: u64,
+    value_size: u64,
+    data: &'data [u8],
+    offsets: &'data [u64],
+}
+
+/// If a column requires allocating a linearized buffer (i.e., when passing a
+/// Vec or slice of AsRef<str> values), this structure is used to hold the
+/// allocated buffers.
+pub struct AllocatedColumn {
+    num_values: u64,
+    value_size: u64,
+    data: Box<[u8]>,
```

Comment:
Not sure which copy you're considering here?

The main reason I went for `Box<[u8]>` is that I was attempting to find a way to just create a large fixed sized buffer in Rust and that was the primary approach I saw while Googling.

However, in hindsight, there are two other decent reasons for preferring it over Vec. First, Vec can be resized which changes it's pointers and I'm not certain that the borrow checker can handle that or not. And if it can't, that seems like a prime recipe for segfaulting when TileDB attempts to use a pointer from a Vec that got resized and invalidated the pointer given to TileDB.

Secondly, probably much less important, but Vec has the size vs capacity issue where a lot of Vec's can end up wasting a lot of RAM due to their memory growth. Its not an issue most of the time, but when it is an issue, it can be a pain to fix.

Reply:
I'm thinking about the `string_to_buffers` function principally, so I'll move the comment over there.

In general, the usual rules of borrow checking apply to `Vec` as to anything else - if something has a reference to the interior data, then nothing is allowed to get a mutable reference to modify it.